internal/storage/inmemory/inmemory.go

```diff
@@ -0,0 +1,202 @@
+package inmemory
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/timuraipov/alert/internal/config"
+	"github.com/timuraipov/alert/internal/domain/metric"
+	"github.com/timuraipov/alert/internal/filestorage"
+	"github.com/timuraipov/alert/internal/logger"
+	"github.com/timuraipov/alert/internal/storage"
+	"go.uber.org/zap"
+)
+
+type InMemory struct {
+	mx              sync.RWMutex
+	fileStorage     *filestorage.Storage
+	DBGauge         map[string]float64
+	DBCounter       map[string]int64
+	config          *config.Config
+	IsNeedSyncFlush bool
+}
+
+func (i *InMemory) init() error {
+	op := "Inmemory.init"
+	if i.config.Restore {
+		err := i.load()
+		if err != nil {
+			logger.Log.Error("can't load file",
+				zap.String("operation", op),
+				zap.Error(err),
+			)
+		}
+	}
+	if i.config.StoreInterval == 0 {
+		i.IsNeedSyncFlush = true
+	} else {
+		tickerFlushToDisk := time.NewTicker(time.Duration(i.config.StoreInterval) * time.Second)
+		done := make(chan struct{})
+		go func() {
+			for {
+				select {
+				case <-tickerFlushToDisk.C:
+					i.Flush()
+				case <-done:
+					tickerFlushToDisk.Stop()
+					return
+				}
+			}
+		}()
+	}
+	return nil
+}
+func (i *InMemory) Flush() error {
+	logger.Log.Debug(
+		"called flush to disk method", zap.String("op", "op"),
+	)
+	metrics, err := i.GetAll(context.Background())
+	if err != nil {
+		return err
+	}
+	if len(metrics) > 0 {
+		data, err := json.Marshal(metrics)
+		if err != nil {
+			return fmt.Errorf("failed to Marshal metrics %w", err)
+		}
+		err = i.fileStorage.Write(data)
+		if err != nil {
+			return fmt.Errorf("failed write metrics to disk %w", err)
+		}
+	}
+	return nil
+}
+func (i *InMemory) load() error {
+	metrics := make([]metric.Metrics, 0)
+	data, err := i.fileStorage.Read()
+	if err != nil {
+		logger.Log.Error("failed to load file", zap.Error(err), zap.String("data", string(data)))
+	}
+	err = json.Unmarshal(data, &metrics)
+	if err != nil {
+		return fmt.Errorf("failed to load file %w", err)
+	}
+	for _, parsedMetric := range metrics {
+		_, err = i.Save(context.Background(), parsedMetric)
+		if err != nil {
+			logger.Log.Error("failed to save metric", zap.Error(err))
+		}
+	}
+	return nil
+}
+
+func (i *InMemory) Save(ctx context.Context, metricObj metric.Metrics) (metric.Metrics, error) {
+	responseList, err := i.save([]metric.Metrics{metricObj})
+	if err != nil {
+		return metric.Metrics{}, err
+	}
+	return responseList[0], nil
+}
+func (i *InMemory) GetAll(ctx context.Context) ([]metric.Metrics, error) {
+	i.mx.RLock()
+	defer i.mx.RUnlock()
+	var metrics []metric.Metrics
```

Comment:
```suggestion
	metrics := make([]metric.Metrics, 0, len(i.DBGauge) + len(i.DBCounter))
```
https://t.me/bushigo/12